config/v1/types_infrastructure.go

```diff
@@ -230,6 +230,36 @@ const (
 	IBMCloudProviderTypeUPI IBMCloudProviderType = "UPI"
 )
 
+// CloudControllerManagerState defines whether Cloud Controller Manager presence is expected or not
+type CloudControllerManagerState string
+
+const (
+	// Cloud Controller Manager is enabled and expected to be installed.
+	// This value indicates that new nodes should be tainted as uninitialized when created,
+	// preventing them from running workloads until they are initialized by the cloud controller manager.
+	CloudControllerManagerExternal CloudControllerManagerState = "External"
+
+	// Cloud Controller Manager is disabled and not expected to be installed.
+	// This value indicates that new nodes should not be tainted
+	// and no extra node initialization is expected from the cloud controller manager.
+	CloudControllerManagerNone CloudControllerManagerState = "None"
+)
+
+// CloudControllerManagerSpec holds Cloud Controller Manager (a.k.a. CCM or CPI) related settings
+type CloudControllerManagerSpec struct {
+	// state determines whether or not an external Cloud Controller Manager is expected to
+	// be installed within the cluster.
+	// https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager
+	//
+	// When set to "External", new nodes will be tainted as uninitialized when created,
+	// preventing them from running workloads until they are initialized by the cloud controller manager.
+	// When omitted or set to "None", new nodes will be not tainted
+	// and no extra initialization from the cloud controller manager is expected.
+	// +kubebuilder:validation:Enum="";External;None
+	// +optional
+	State CloudControllerManagerState `json:"state"`
```

Comment:
Do we allow this to be changed post install? Are there any transitions we would like to prevent, eg, should the `External` not be allowed to transition back to `None`?

Reply:
we have never discussed being able to transit back from CCMs to non-CCMs. given that this field is intended for deployments where there most likely won't be an in-tree provider, i don't think we will allow it to be changed after install.

i'm trying to think about what that would mean, and essentially it would be switching a cluster from using CCMs to using no cloud controllers. which sounds like it might work in theory, but i'm not sure about the transitive effects and if that is supportable for us.

Reply:
Joel and i discussed this in our standup today, we have some questions about whether this field should be in the spec or status of the infra object. i'd like to get a few more opinions about how we intend this feature to be used, and if it is appropriate for users to change it after installation.